### PR TITLE
Purego

### PR DESCRIPTION
**Pull request by gysddn**
This is a somewhat aggressive refactor of the current `cgo` bridge. 

So far, most, almost all, functions calls have been integrated to `purego`.
If the end goal is to get rid of `cgo` altogether, this is the todo list:

- [x]  Locate library files dynamically.
- [x] Make sure the strings passed to purego are [null terminated](https://pkg.go.dev/github.com/jwijenbergh/purego#hdr-Memory-RegisterFunc).
- [x] Handle memory of strings in general
- [x] Handle callbacks in purego.
- [x] Go through every C.* call and update them to purego (i.e C.CString).
- [x] Remove cgo types
- [x] Appindicator

Testing Window API (Manually)
- [x] Load
- [x] Unload (Works but throws an error when called in a `goroutine`.)
- [x] Focus (`gtk_widget_grab_focus` documents using this to give a widget the keyboard focus, and API calls it on a window?)
- [x] SetVisible
- [x] IsVisible
- [x] SetMaximized
- [x] SetMinimized (Works but I couldn't test `false` condition since I'm on GNOME and it does not respond to `deiconify`.)
- [x] SetFullscreen
- [x] SetSize
- [x] SetMinSize and SetMaxSize (Even though API separates these, gtk_window_set_geometry_hints ties them together and in my tests it seems only the last call takes effect. This can also be seen [here](https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gtk/gtkwindow.c#L4098). So this is TODO item)
- [x] SetResizable
- [x] SetAlwaysOnTop
- [x] SetPosition
- [x] SetTitle
- [x] GetOuterPosition (There has to be a small delay between the set and the get, otherwise it returns the old position.)
- [x] GetOuterSize
- [x] GetInnerSize

### CODE REVIEW

**Comment by gysddn** [comment]
About:

> - [ ] Remove cgo types

If removing `cgo` is not a very strict requirement, I'd like to point out that it provides the types which in turn allows type checking on pointers, that wouldn't otherwise be possible (to my knowledge) in a `purego` only configuration.

I can, of course, convert them into `purego` only calls but I wanted to ask for an input about this @progrium before doing so.

Purego with cgo:

```go
    GtkWidgetShowAll                          func (widget *C.GtkWidget)
```

Purego only:

```go
    GtkWidgetShowAll                          func (widget unsafe.Pointer)
```

**Comment by progrium** [comment]
> About:
> 
> > * [ ]  Remove cgo types
> 
> If removing `cgo` is not a very strict requirement, I'd like to point out that it provides the types which in turn allows type checking on pointers, that wouldn't otherwise be possible (to my knowledge) in a `purego` only configuration.
> 
> I can, of course, convert them into `purego` only calls but I wanted to ask for an input about this @progrium before doing so.
> 
> Purego with cgo:
> 
> ```go
>     GtkWidgetShowAll                          func (widget *C.GtkWidget)
> ```
> 
> Purego only:
> 
> ```go
>     GtkWidgetShowAll                          func (widget unsafe.Pointer)
> ```

Thanks for asking for clarification here. We do have a hard requirement to remove CGO so I am ok with `unsafe.Pointer` internally. If its exposed to the user, we'll do something extra, but I'm pretty sure it's not. Is that correct?

**Comment by gysddn** [comment]
Last commit should be "GdkRectangle" not "GdkGeometry", my bad there 😅
Don't want to force-push while in review

**Comment by gysddn** [comment]
> Thanks for asking for clarification here. We do have a hard requirement to remove CGO so I am ok with `unsafe.Pointer` internally. If its exposed to the user, we'll do something extra, but I'm pretty sure it's not. Is that correct?

Correct. Only exposed types are structs that hold pointers to those types, which are hidden away, so changing them didn't break anything.

**Comment by gysddn** [comment]
I have tested the API after the last change and reported the problems I found as issues. Other than that, the PR is ready for review. @progrium

File: desktop/linux/linux.go [hunk 1]
```diff
@@ -8,55 +8,101 @@ import (
...
+	Y int32
 }
 
 type Rectangle struct {
 	Position Position
 	Size     Size
 }
 
-type EventType int
+type GdkGeometry struct {
```
Comment (progrium) [inline desktop/linux/linux.go:56]: Even though these represent C structs, we can name fields whatever we want. Let's use Go naming conventions.

File: desktop/linux/linux.go [hunk 2]
```diff
@@ -93,27 +139,501 @@ type WebviewSetings struct {
 	DeveloperTools       bool
 }
 
+// GtkWindowType
+const (
+	GTK_WINDOW_TOPLEVEL = iota
```
Comment (progrium) [inline desktop/linux/linux.go:144]: Also for constants, let's keep using Go naming conventions (no all caps, no underscores, ...)

File: desktop/linux/linux.go [hunk 3]
```diff
@@ -93,27 +139,501 @@ type WebviewSetings struct {
...
+		data unsafe.Pointer,
+		//These two last arguments are ignored, so they only have generic types here
+		destroy_data unsafe.Pointer,
+		connect_flags uint32,
+	)
+)
+
+var (
+	//TODO put these in lexical order
+	//TODO indentation
```
Comment (progrium) [inline desktop/linux/linux.go:291]: yes lets clean up indentation

**Review by progrium — CHANGES_REQUESTED** [review]
Looks good just some cosmetic changes and then I'll merge.

**Comment by gysddn** [comment]
Also added function comment docs to denote the actual types of unsafe.Pointers, and to document functions in general.

**Comment by gysddn** [comment]
@progrium I have cleaned up the file and formatted with proper indentation.
PR is ready for review.

I have a commit for app indicators and working on the handling of `SetSize` (#8) which is partially ready. But I didn't want to make this PR more crowded so I'm planning to submit them as separate PRs after this one is merged.

**Pull request merged**